Precinct Selection Scan Fix + Refactor

adghayes opened the pull request:
Fixes bug on main where no ballots will scan in precinct scanner on all precincts mode: https://votingworks.slack.com/archives/CEL6D3GAD/p1661984695650299.

Standardizes the `PrecinctSelection` type across apps, and refactors in other apps where appropriate. I recommend reviewing by app:
- `libs/types`: some additions and edits to to `PrecinctSelection` types
- `libs/ui`: use type in `ElectionInfoBar`
- `services/scan`: use and store `PrecinctSelection` rather than `PrecinctId`
- `frontends/precinct-scanner`: use `PrecinctSelection` rather than `PrecinctId`
- `frontends/bmd`: use type from `libs/types` rather than the frontend types

arsalansufi commented on libs/types/src/precinct_selection.ts:
```diff
@@ -1,22 +1,65 @@
+
+export const ALL_PRECINCTS_NAME = 'All Precincts';
+
+export function getPrecinctSelectionName(
```
This might actually better fit in `libs/utils` - if we put it there, we can also make us of the `find` util instead of manually recreating that logic. The same could be said of `getSinglePrecinctSelection`

adghayes replied:
Will do!

arsalansufi commented on services/scan/src/importer.ts:
```diff
@@ -136,6 +137,8 @@ export class Importer {
   configure(electionDefinition: ElectionDefinition): void {
     this.workspace.store.setElection(electionDefinition);
+    // Central scanner only uses all precinct mode, set on every configure
+    this.workspace.store.setPrecinctSelection(ALL_PRECINCTS_SELECTION);
```
Since this added logic is specific to the central scanner, we may want to move it up to `services/scan/src/central_scanner_app.ts` instead of keeping it in this lower, machine-agnostic layer

adghayes replied:
I originally had it in the higher layer but it had to be in a few places - both on initial setup and any reconfigures - to work properly.

The main point to make is that, I believe, `importer.ts` is not actually machine-agnostic. It is only used for VxCentralScan. I think @jonahkagan told me this. Will let him or @eventualbuddha confirm.

adghayes replied:
I was confused by this at first and if `importer.ts` is machine specific now, we might rename the file to reflect that.

arsalansufi replied:
Ah interesting - I like the idea of renaming then! Maybe `central_scanner_importer.ts` / `CentralScanImporter`

adghayes replied:
https://github.com/votingworks/vxsuite/issues/2532

jonahkagan replied:
Yep, importer is now only used by the central scanner. Renaming it makes sense!

arsalansufi commented on services/scan/src/store.ts:
```diff
@@ -57,7 +58,7 @@ export enum ConfigKey {
   // @deprecated
   SkipElectionHashCheck = 'skipElectionHashCheck',
-  CurrentPrecinctId = 'currentPrecinctId',
+  PrecinctSelectionKey = 'precinctSelection',
```
For consistency with the other keys in this enum, we may want to shorten this to `PrecinctSelection = 'precinctSelection'`. The `Key` in `ConfigKey` should make it clear that this is a key, e.g. `ConfigKey.PrecinctSelection`

adghayes replied:
I tried this but there's a variable name conflict in the file if you do that, will take another look to see if there's another way.

adghayes replied:
Renamed the key by importing the type as:
```
  PrecinctSelection as PrecinctSelectionType,
```

arsalansufi commented on services/scan/src/workers/interpret_vx.ts:
```diff
@@ -55,6 +59,7 @@ export async function configure(store: Store): Promise<void> {
   debug('creating a new interpreter');
   interpreter = new Interpreter({
     electionDefinition,
+    precinctSelection: ALL_PRECINCTS_SELECTION,
```
Is there a reason why we're using `ALL_PRECINCTS_SELECTION` here instead of `store.getPrecinctSelection()`?

adghayes replied:
This is a really helpful catch @arsalansufi, I think this was a mistake.

arsalansufi replied:
Ah nice!

adghayes replied:
Fixed.

jonahkagan replied:
@adghayes were the tests passing with this mistake? If so, that might be a bit concerning

arsalansufi commented on libs/ui/src/election_info_bar.tsx:
```diff
@@ -48,9 +43,15 @@ export function ElectionInfoBar({
             <React.Fragment>
-              <NoWrap>{precinct?.name ?? 'All Precincts'}</NoWrap> —{' '}
+              <NoWrap>
+                {getPrecinctSelectionName(
```
A slightly more intuitive name for this helper might be `getSelectedPrecinctName` or `precinctSelectionToReadableString`

adghayes replied:
The current name makes the most sense to me, as `Selected` feels like it implies a selection was made when that was not always the case and `ReadableString` is less clear than `Name` only because `Name` is how we've been identifying these strings in code to date.

arsalansufi replied:
Ah good point that the precinct isn't actually selected on VxCentralScan - the current name is okay then!

arsalansufi commented on frontends/bmd/src/pages/admin_screen.tsx:
```diff
@@ -57,12 +59,9 @@ export function AdminScreen({
-        kind: PrecinctSelectionKind.SinglePrecinct,
-        precinctId,
-      });
+      updateAppPrecinct(getSinglePrecinctSelection(precinctId));
```
I don't feel strongly about this! `constructSinglePrecinctSelection` might be a clearer name than `getSinglePrecinctSelection`, which reads as if we're retrieving something from a store

adghayes replied:
Renamed to `singlePrecinctSelectionFor`.

arsalansufi reviewed (commented):
Thank you for making this consistent across VxSuite! Just a few thoughts!

arsalansufi reviewed (approved):
LGTM!

jonahkagan commented on services/scan/src/central_scanner_app.ts:
```diff
@@ -197,21 +197,21 @@ export async function buildCentralScannerApp({
   });
 
-  app.get<NoParams, Scan.GetCurrentPrecinctConfigResponse>(
+  app.get<NoParams, Scan.GetPrecinctSelectionConfigResponse>(
```
Does the central scanner use these endpoints at all? As far as I can tell, no. And I don't see why it would need them since it's not configurable by precinct. They can probably be removed (along with the corresponding functions in frontends/bsd/src/api/config.ts)

adghayes replied:
No, we don't use them, wasn't sure if I should remove them. I will go ahead and do so.

adghayes replied:
Removed.

jonahkagan commented on services/scan/src/importer.ts:
```diff
@@ -236,17 +241,18 @@ export class Importer {
-    debug('currentPrecinctId=%s', currentPrecinctId);
-    if (currentPrecinctId) {
+    debug('precinctSelection=%o', precinctSelection);
+    if (precinctSelection.kind !== 'AllPrecincts') {
```
Isn't this case now impossible?

adghayes replied:
Yes, this is a place where we still have precinct-specific logic in the central scanner. Like with the endpoint, we can remove them.

adghayes replied:
Removed.

jonahkagan commented on services/scan/src/precinct_scanner_app.ts:
```diff
@@ -248,15 +250,15 @@ export async function buildPrecinctScannerApp(
   });
 
-  app.delete<NoParams, Scan.DeleteCurrentPrecinctConfigResponse>(
+  app.delete<NoParams, Scan.DeletePrecinctSelectionConfigResponse>(
```
Is this endpoint used anymore? I noticed you removed it from the tests

adghayes replied:
It is being used, yes, I'll have to check where I removed it from the tests.

jonahkagan replied:
I guess a more relevant question would be: is there a use case for clearing the VxScan precinct selection back to an unconfigured state? I can't really see why a user would want to do that personally.

adghayes replied:
When you unconfigure the machine, you want it to return to a no precinct state. The precinct selection makes no sense outside of the context of the election definition, so if the election definition is removed, the precinct selection should be too.

Are you suggesting that we delete this endpoint and then just remove the precinct selection on zeroing out the scanner? That makes more sense to me, now that I'm thinking about it.

jonahkagan replied:
Yeah, and I think that's actually what we already do. Just left another comment about it haha

jonahkagan commented on frontends/precinct-scanner/src/screens/election_manager_screen.tsx:
```diff
@@ -126,6 +137,12 @@ export function ElectionManagerScreen({
+    ? precinctSelection.kind === 'AllPrecincts'
+      ? ALL_PRECINCTS_OPTION_VALUE
+      : precinctSelection.precinctId
+    : '';
```
This is a bit of a confusing name. The `Select` is a controlled component, so its value will always be whatever is passed into the `value` prop. So this variable isn't the "initial" value but rather just the value. I might name it `precinctSelectionValue` or something like that

adghayes replied:
Renamed.

jonahkagan commented on frontends/bmd/src/pages/admin_screen.tsx:
```diff
@@ -103,7 +101,7 @@ export function AdminScreen({
                   </option>
                   <option value={ALL_PRECINCTS_OPTION_VALUE}>
-                    {AllPrecinctsDisplayName}
+                    All Precincts
```
Didn't you make a constant for this in the precinct selection utils?

adghayes replied:
Yes, I just didn't bother importing it here. Can do so.

adghayes replied:
Switched.

jonahkagan reviewed (approved):
Nice work! The refactoring makes things a lot cleaner. Left a few other ideas for cleaning up more. If possible, it would make it easier to review if you can separate out the refactoring into a separate PR next time (or separate commits).

jonahkagan commented on services/scan/src/precinct_scanner_app.test.ts:
```diff
@@ -120,6 +119,14 @@ function postTemplate(
+      ? singlePrecinctSelectionFor(precinctId)
+      : ALL_PRECINCTS_SELECTION,
+  });
+}
```
I realized we don't have a test that confirms that unconfiguring the scanner using the `DELETE /precinct-scanner/config/election` endpoint also clears the precinct config. Reading the code, I _think_ it works, but might be worth checking/testing if you haven't already, since the frontend doesn't explicitly clear the precinct config on unconfigure.

adghayes replied:
I don't think we have a test that covers the` DELETE /precinct-scanner/config/election` endpoint at all. Will do some manual testing.

adghayes replied:
We should also of course write a test, but I need to close this out and need to do manual testing anyway.

adghayes commented:
@jonahkagan I addressed most of your feedback. In manual testing, I found that my changes had broken the configuration process for VxScan. It was because `configureInterpreter` now requires a precinct selection, but was being called without one having been made. I changed the logic of when we configure the interpreter in `precinct_scanner_app.ts`. The central scanner hits the `/doneTemplates` endpoint after all templates have been uploaded which triggers a configure. That's what we were doing in `precinct-scanner`, but now that you have to select a precinct, you can't configure the interpreter until after a precinct selection. All automated testing was passing before I made this change.

To your previous question, automated testing was passing before I fixed the mistake that @arsalansufi pointed out. I have not dug into why that was the case yet. 

I'm not going to merge this until I get a LGTM or another close look on your part. This turned out, in general, to be a much deeper dive than expected.

jonahkagan commented on services/scan/src/precinct_scanner_app.ts:
```diff
@@ -91,7 +91,7 @@ export async function buildPrecinctScannerApp(
 
   // Load interpreter configuration from the store on startup
-  if (store.getElectionDefinition()) {
+  if (store.getElectionDefinition() && store.getPrecinctSelection()) {
```
Since there are now a bunch of repeated checks that the precinct selection is configured before configuring the interpreter, we could move them inside the `configureInterpreter` function (along with the check for election definition).

adghayes replied:
I like that, will do one last refactor. Seems like a potential foot gun later if we don't.

adghayes commented:
Did some manual testing on VxScan (testing different precinct configurations, and reconfigure) and VxCentralScan (testing scanning, and reconfigure). Looks good. Merging.
